File: test/range_sum_of_bst.js
const chai = require('chai');
const expect = chai.expect;
const tree = require('leetcode').Tree;

const rangeSumBST = (node, L, R) => {
  let total = 0;
  const traverse = (node, L, R) => {
    if(!node) return;
    if(node.val >= L && node.val <= R) total += node.val;
    if(node.left) traverse(node.left, L, R);
    if(node.right) traverse(node.right, L, R);
  };
  traverse(node, L, R);
  return total;
};

describe('Range sum of BST', () => {
  it('Example 1', () => {
    let root = [10, 5, 15, 3, 7, null, 18], L = 7, R = 15;
    let node = tree.create(root);
    let actual = rangeSumBST(node, L, R);
    let expected = 32;
    expect(actual).to.equal(expected);
  });
  it('Example 2', () => {
    let root = [10, 5, 15, 3, 7, 13, 18, 1, null, 6], L = 6, R = 10;
    let node = tree.create(root);
    let actual = rangeSumBST(node, L, R);
    let expected = 23;
    expect(actual).to.equal(expected);
  });
});

File: Perl/t/range_sum_of_bst.pl
#!/usr/bin/env perl
use strict;
use warnings;
use version; our $VERSION = qv('1.0.0');

use Test::More::Behaviour;
use lib "t";
use BTreeNode;

use feature 'current_sub';

sub range_sum_bst {
    my ( $node, $L, $R ) = @_;
    my $total = 0;

    my $traverse = sub {

        my ( $node, $L, $R ) = @_;

        if ( !$node ) { return; }
        if ( $node->{val} >= $L && $node->{val} <= $R ) {
            $total += $node->{val};
        }
        if ( $node->{left} )  { __SUB__->( $node->{left},  $L, $R ) }
        if ( $node->{right} ) { __SUB__->( $node->{right}, $L, $R ) }
    };

    $traverse->( $node, $L, $R );
    return $total;
}

sub assert_range_sum_bst {
    my ( $tree_ref, $left, $right, $expected, $message ) = @_;

    it $message => sub {
        my $root = BTreeNode->deserialize( @{$tree_ref} );
        my $got  = range_sum_bst( $root, $left, $right );
        is( $got, $expected );
    }
}

describe 'Range sum of binary search tree' => sub {
    assert_range_sum_bst( [ 10, 5, 15, 3, 7, undef, 18 ],
        7, 15, 32, "Example 1" );
    assert_range_sum_bst( [ 10, 5, 15, 3, 7, 13, 18, 1, undef, 6 ],
        6, 10, 23, "Example 2" );
};

done_testing;
1;
